File: thrifty-schema/src/test/java/com/microsoft/thrifty/schema/ConstantTest.java
/*
 * Thrifty
 *
 * Copyright (c) Microsoft Corporation
 *
 * All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the License);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING
 * WITHOUT LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE,
 * FITNESS FOR A PARTICULAR PURPOSE, MERCHANTABLITY OR NON-INFRINGEMENT.
 *
 * See the Apache Version 2.0 License for specific language governing permissions and limitations under the License.
 */
package com.microsoft.thrifty.schema;

import com.google.common.collect.ImmutableList;
import com.microsoft.thrifty.schema.parser.ConstElement;
import com.microsoft.thrifty.schema.parser.ConstValueElement;
import com.microsoft.thrifty.schema.parser.EnumElement;
import com.microsoft.thrifty.schema.parser.EnumMemberElement;
import com.microsoft.thrifty.schema.parser.ThriftFileElement;
import com.microsoft.thrifty.schema.parser.ThriftParser;
import com.microsoft.thrifty.schema.parser.TypeElement;
import com.microsoft.thrifty.schema.parser.TypedefElement;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.hamcrest.CoreMatchers.containsString;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.fail;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class ConstantTest {
    @Mock SymbolTable symbolTable;
    private Location loc = Location.get("", "");

    @Test
    public void boolLiteral() {
        Constant.validate(symbolTable, ConstValueElement.identifier(loc, "true", "true"), BuiltinType.BOOL);
        Constant.validate(symbolTable, ConstValueElement.identifier(loc, "false", "false"), BuiltinType.BOOL);
        try {
            Constant.validate(symbolTable, ConstValueElement.literal(loc, "nope", "nope"), BuiltinType.BOOL);
            fail("Invalid identifier should not validate as a bool");
        } catch (IllegalStateException expected) {
            assertThat(
                    expected.getMessage(),
                    containsString("Expected 'true', 'false', '1', '0', or a bool constant"));
        }
    }

    @Test
    public void boolConstant() {
        Constant c = makeConstant(
                "aBool",
                TypeElement.scalar(loc, "string", null),
                ConstValueElement.identifier(loc, "aBool", "aBool"),
                BuiltinType.BOOL);

        when(symbolTable.lookupConst("aBool")).thenReturn(c);

        Constant.validate(symbolTable, ConstValueElement.identifier(loc, "aBool", "aBool"), BuiltinType.BOOL);
    }

    @Test
    public void boolWithWrongTypeOfConstant() {
        Constant c = makeConstant(
                "aBool",
                TypeElement.scalar(loc, "string", null),
                ConstValueElement.identifier(loc, "aBool", "aBool"),
                BuiltinType.STRING);

        when(symbolTable.lookupConst("aBool")).thenReturn(c);

        try {
            Constant.validate(symbolTable, ConstValueElement.identifier(loc, "aBool", "aBool"), BuiltinType.BOOL);
            fail("Wrongly-typed constant should not validate");
        } catch (IllegalStateException ignored) {
        }
    }

    @Test
    public void boolWithNonConstantIdentifier() {
        try {
            Constant.validate(symbolTable, ConstValueElement.identifier(loc, "someStruct", "someStruct"), BuiltinType.BOOL);
            fail("Non-constant identifier should not validate");
        } catch (IllegalStateException expected) {
            assertThat(
                    expected.getMessage(),
                    containsString("Expected 'true', 'false', '1', '0', or a bool constant; got: someStruct"));
        }
    }

    @Test
    public void boolWithConstantHavingBoolTypedefValue() {
        TypedefType td = makeTypedef(BuiltinType.BOOL, "Truthiness");

        Constant c = makeConstant(
                "aBool",
                TypeElement.scalar(loc, "Truthiness", null),
                ConstValueElement.identifier(loc, "aBool", "aBool"),
                td);

        when(symbolTable.lookupConst("aBool")).thenReturn(c);

        Constant.validate(symbolTable, ConstValueElement.identifier(loc, "aBool", "aBool"), BuiltinType.BOOL);
    }

    @Test
    public void typedefWithCorrectLiteral() {
        TypedefType td = makeTypedef(BuiltinType.STRING, "Message");

        ConstValueElement value = ConstValueElement.literal(loc, "\"y helo thar\"", "y helo thar");

        Constant.validate(symbolTable, value, td);
    }

    @Test
    public void inRangeInt() {
        ConstValueElement value = ConstValueElement.integer(loc, "10", 10);
        ThriftType type = BuiltinType.I32;

        Constant.validate(symbolTable, value, type);
    }

    @Test
    public void tooLargeInt() {
        ConstValueElement value = ConstValueElement.integer(
                loc, String.valueOf((long) Integer.MAX_VALUE + 1), (long) Integer.MAX_VALUE + 1);
        ThriftType type = BuiltinType.I32;

        try {
            Constant.validate(symbolTable, value, type);
            fail("out-of-range i32 should fail");
        } catch (IllegalStateException e) {
            assertThat(e.getMessage(), containsString("out of range for type i32"));
        }
    }

    @Test
    public void tooSmallInt() {
        ConstValueElement value = ConstValueElement.integer(
                loc, String.valueOf((long) Integer.MIN_VALUE - 1), (long) Integer.MIN_VALUE - 1);
        ThriftType type = BuiltinType.I32;

        try {
            Constant.validate(symbolTable, value, type);
            fail("out-of-range i32 should fail");
        } catch (IllegalStateException e) {
            assertThat(e.getMessage(), containsString("out of range for type i32"));
        }
    }

    @Test
    public void enumWithMember() {
        List<EnumMemberElement> memberElements = Collections.singletonList(
                new EnumMemberElement(loc, "TEST", 1)
        );

        EnumElement enumElement = new EnumElement(loc, "TestEnum", memberElements);

        EnumType et = new EnumType(enumElement, Collections.emptyMap());

        Constant.validate(symbolTable, ConstValueElement.identifier(loc, "TestEnum.TEST", "TestEnum.TEST"), et);
    }

    @Test
    public void enumWithNonMemberIdentifier() {
        List<EnumMemberElement> memberElements = Collections.singletonList(
                new EnumMemberElement(loc, "TEST", 1)
        );

        EnumElement enumElement = new EnumElement(loc, "TestEnum", memberElements);

        EnumType et = new EnumType(enumElement, Collections.emptyMap());

        try {
            Constant.validate(symbolTable, ConstValueElement.identifier(loc, "TestEnum.NON_MEMBER", "TestEnum.NON_MEMBER"), et);
            fail("Non-member identifier should fail");
        } catch (IllegalStateException expected) {
            assertThat(
                    expected.getMessage(),
                    containsString("'TestEnum.NON_MEMBER' is not a member of enum type TestEnum: members=[TEST]"));
        }
    }

    @Test
    public void unqualifiedEnumMember() {
        List<EnumMemberElement> memberElements = Collections.singletonList(
                new EnumMemberElement(loc, "TEST", 1)
        );

        EnumElement enumElement = new EnumElement(loc, "TestEnum", memberElements);

        EnumType et = new EnumType(enumElement, Collections.emptyMap());

        try {
            Constant.validate(symbolTable, ConstValueElement.identifier(loc, "TEST", "TEST"), et);
            fail("Expected an IllegalStateException");
        } catch (IllegalStateException e) {
            assertThat(e.getMessage(), containsString("Unqualified name 'TEST' is not a valid enum constant value"));
        }
    }

    @Test
    public void listOfInts() {
        ThriftType list = new ListType(BuiltinType.I32);
        ConstValueElement listValue = ConstValueElement.list(loc, "[0, 1, 2]", Arrays.asList(
                ConstValueElement.integer(loc, "0", 0),
                ConstValueElement.integer(loc, "1", 1),
                ConstValueElement.integer(loc, "2", 2)
        ));

        Constant.validate(symbolTable, listValue, list);
    }

    @Test
    public void heterogeneousList() {
        ThriftType list = new ListType(BuiltinType.I32);
        ConstValueElement listValue = ConstValueElement.list(loc, "[0, 1, \"2\"]", Arrays.asList(
                ConstValueElement.integer(loc, "0", 0),
                ConstValueElement.integer(loc, "1", 1),
                ConstValueElement.literal(loc, "\"2\"", "2")
        ));

        try {
            Constant.validate(symbolTable, listValue, list);
            fail("Heterogeneous lists should fail validation");
        } catch (IllegalStateException ignored) {
        }
    }

    @Test
    public void typedefOfEnum() {
        List<EnumMemberElement> memberElements = Collections.singletonList(
                new EnumMemberElement(loc, "FOO", 1)
        );

        EnumElement enumElement = new EnumElement(loc, "AnEnum", memberElements);

        EnumType et = new EnumType(enumElement, Collections.emptyMap());

        TypedefType typedefType = makeTypedef(et, "Id");

        ConstValueElement value = ConstValueElement.identifier(loc, "AnEnum.FOO", "AnEnum.FOO");

        Constant.validate(symbolTable, value, typedefType);
    }

    @Test
    public void typedefOfWrongEnum() {
        List<EnumMemberElement> memberElements = Collections.singletonList(
                new EnumMemberElement(loc, "FOO", 1)
        );

        EnumElement enumElement = new EnumElement(loc, "AnEnum", memberElements);

        EnumType et = new EnumType(enumElement, Collections.emptyMap());

        List<EnumMemberElement> wrongMemberElements = Collections.singletonList(
                new EnumMemberElement(loc, "BAR", 2)
        );

        EnumElement wrongEnumElement = new EnumElement(loc, "DifferentEnum", wrongMemberElements);

        EnumType wt = new EnumType(wrongEnumElement, Collections.emptyMap());

        TypedefType typedefType = makeTypedef(wt, "Id");

        ConstValueElement value = ConstValueElement.identifier(loc, "AnEnum.FOO", "AnEnum.FOO");

        try {
            Constant.validate(symbolTable, value, typedefType);
            fail("An enum literal of type A cannot be assigned to a typedef of type B");
        } catch (IllegalStateException expected) {
            assertThat(
                    expected.getMessage(),
                    is("'AnEnum.FOO' is not a member of enum type DifferentEnum: members=[BAR]"));
        }
    }

    @Test
    public void setOfInts() {
        ThriftType setType = new SetType(BuiltinType.I32);
        ConstValueElement listValue = ConstValueElement.list(loc, "[0, 1, 2]", Arrays.asList(
                ConstValueElement.integer(loc, "0", 0),
                ConstValueElement.integer(loc, "1", 1),
                ConstValueElement.integer(loc, "2", 2)
        ));

        Constant.validate(symbolTable, listValue, setType);
    }

    private Constant makeConstant(String name, TypeElement typeElement, ConstValueElement value, ThriftType thriftType) {
        ConstElement element = new ConstElement(
                loc,
                typeElement,
                name,
                value);

        try {
            Constant constant = new Constant(element, Collections.emptyMap());

            Field field = Constant.class.getDeclaredField("type");
            field.setAccessible(true);
            field.set(constant, thriftType);

            return constant;
        } catch (NoSuchFieldException | IllegalAccessException e) {
            throw new AssertionError(e);
        }
    }

    private TypedefType makeTypedef(ThriftType oldType, String newName) {
        TypedefElement element = new TypedefElement(
                loc,
                TypeElement.scalar(loc, "does_not_matter", null),
                newName);

        try {
            TypedefType td = new TypedefType(Collections.emptyMap(), element);

            Field field = TypedefType.class.getDeclaredField("oldType");
            field.setAccessible(true);
            field.set(td, oldType);

            return td;
        } catch (IllegalAccessException | NoSuchFieldException e) {
            throw new AssertionError(e);
        }
    }
}

File: thrifty-schema/src/test/kotlin/com/microsoft/thrifty/schema/ConstantTest.kt
/*
 * Thrifty
 *
 * Copyright (c) Microsoft Corporation
 *
 * All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the License);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING
 * WITHOUT LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE,
 * FITNESS FOR A PARTICULAR PURPOSE, MERCHANTABLITY OR NON-INFRINGEMENT.
 *
 * See the Apache Version 2.0 License for specific language governing permissions and limitations under the License.
 */
package com.microsoft.thrifty.schema

import com.microsoft.thrifty.schema.parser.ConstElement
import com.microsoft.thrifty.schema.parser.ConstValueElement
import com.microsoft.thrifty.schema.parser.DoubleValueElement
import com.microsoft.thrifty.schema.parser.EnumElement
import com.microsoft.thrifty.schema.parser.EnumMemberElement
import com.microsoft.thrifty.schema.parser.IdentifierValueElement
import com.microsoft.thrifty.schema.parser.IntValueElement
import com.microsoft.thrifty.schema.parser.ListValueElement
import com.microsoft.thrifty.schema.parser.LiteralValueElement
import com.microsoft.thrifty.schema.parser.ScalarTypeElement
import com.microsoft.thrifty.schema.parser.TypeElement
import com.microsoft.thrifty.schema.parser.TypedefElement
import org.junit.Test

import org.hamcrest.CoreMatchers.containsString
import org.hamcrest.Matchers.`is`
import org.junit.Assert.assertThat
import org.junit.Assert.fail
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`

class ConstantTest {
    private val symbolTable: SymbolTable = mock(SymbolTable::class.java)
    private val loc = Location.get("", "")

    @Test
    fun boolLiteral() {
        Constant.validate(symbolTable, IdentifierValueElement(loc, "true", "true"), BuiltinType.BOOL)
        Constant.validate(symbolTable, IdentifierValueElement(loc, "false", "false"), BuiltinType.BOOL)
        try {
            Constant.validate(symbolTable, LiteralValueElement(loc, "nope", "nope"), BuiltinType.BOOL)
            fail("Invalid identifier should not validate as a bool")
        } catch (expected: IllegalStateException) {
            assertThat<String>(
                    expected.message,
                    containsString("Expected 'true', 'false', '1', '0', or a bool constant"))
        }
    }

    @Test
    fun boolConstant() {
        val c = makeConstant(
                "aBool",
                ScalarTypeElement(loc, "string", null),
                IdentifierValueElement(loc, "aBool", "aBool"),
                BuiltinType.BOOL)

        `when`<Constant>(symbolTable.lookupConst("aBool")).thenReturn(c)

        Constant.validate(symbolTable, IdentifierValueElement(loc, "aBool", "aBool"), BuiltinType.BOOL)
    }

    @Test
    fun boolWithWrongTypeOfConstant() {
        val c = makeConstant(
                "aBool",
                ScalarTypeElement(loc, "string", null),
                IdentifierValueElement(loc, "aBool", "abool"),
                BuiltinType.STRING)

        `when`<Constant>(symbolTable.lookupConst("aBool")).thenReturn(c)

        try {
            Constant.validate(symbolTable, IdentifierValueElement(loc, "aBool", "aBool"), BuiltinType.BOOL)
            fail("Wrongly-typed constant should not validate")
        } catch (ignored: IllegalStateException) {
        }

    }

    @Test
    fun boolWithNonConstantIdentifier() {
        try {
            Constant.validate(symbolTable, IdentifierValueElement(loc, "someStruct", "someStruct"), BuiltinType.BOOL)
            fail("Non-constant identifier should not validate")
        } catch (expected: IllegalStateException) {
            assertThat<String>(
                    expected.message,
                    containsString("Expected 'true', 'false', '1', '0', or a bool constant; got: someStruct"))
        }

    }

    @Test
    fun boolWithConstantHavingBoolTypedefValue() {
        val td = makeTypedef(BuiltinType.BOOL, "Truthiness")

        val c = makeConstant(
                "aBool",
                ScalarTypeElement(loc, "Truthiness", null),
                IdentifierValueElement(loc, "aBool", "aBool"),
                td)

        `when`<Constant>(symbolTable.lookupConst("aBool")).thenReturn(c)

        Constant.validate(symbolTable, IdentifierValueElement(loc, "aBool", "aBool"), BuiltinType.BOOL)
    }

    @Test
    fun typedefWithCorrectLiteral() {
        val td = makeTypedef(BuiltinType.STRING, "Message")

        val value = LiteralValueElement(loc, "\"y helo thar\"", "y helo thar")

        Constant.validate(symbolTable, value, td)
    }

    @Test
    fun inRangeInt() {
        val value = IntValueElement(loc, "10", 10)
        val type = BuiltinType.I32

        Constant.validate(symbolTable, value, type)
    }

    @Test
    fun tooLargeInt() {
        val value = IntValueElement(loc,
                "${Integer.MAX_VALUE.toLong() + 1}", Integer.MAX_VALUE.toLong() + 1)
        val type = BuiltinType.I32

        try {
            Constant.validate(symbolTable, value, type)
            fail("out-of-range i32 should fail")
        } catch (e: IllegalStateException) {
            assertThat<String>(e.message, containsString("out of range for type i32"))
        }

    }

    @Test
    fun tooSmallInt() {
        val value = IntValueElement(loc,
                "${Integer.MIN_VALUE.toLong() - 1}", Integer.MIN_VALUE.toLong() - 1)
        val type = BuiltinType.I32

        try {
            Constant.validate(symbolTable, value, type)
            fail("out-of-range i32 should fail")
        } catch (e: IllegalStateException) {
            assertThat<String>(e.message, containsString("out of range for type i32"))
        }

    }

    @Test
    fun doubleLiteral() {
        Constant.validate(symbolTable, IntValueElement(loc, "10", 10), BuiltinType.DOUBLE)
        Constant.validate(symbolTable, DoubleValueElement(loc, "3.14", 3.14), BuiltinType.DOUBLE)
        try {
            Constant.validate(symbolTable, LiteralValueElement(loc, "aString", "aString"), BuiltinType.DOUBLE)
            fail("String literal should not validate as a double")
        } catch (expected: IllegalStateException) {
            assertThat<String>(
                    expected.message,
                    containsString("Expected a value of type DOUBLE but got aString"))
        }
    }

    @Test
    fun doubleConstant() {
        val c = makeConstant(
                "aDouble",
                ScalarTypeElement(loc, "string", null),
                IdentifierValueElement(loc, "aDouble", "aDouble"),
                BuiltinType.DOUBLE)

        `when`<Constant>(symbolTable.lookupConst("aDouble")).thenReturn(c)

        Constant.validate(symbolTable, IdentifierValueElement(loc, "aDouble", "aDouble"), BuiltinType.DOUBLE)
    }

    @Test
    fun doubleWithWrongTypeOfConstant() {
        val c = makeConstant(
                "aString",
                ScalarTypeElement(loc, "string", null),
                IdentifierValueElement(loc, "aString", "aString"),
                BuiltinType.STRING)

        `when`<Constant>(symbolTable.lookupConst("aString")).thenReturn(c)

        try {
            Constant.validate(symbolTable, IdentifierValueElement(loc, "aString", "aString"), BuiltinType.DOUBLE)
            fail("Wrongly-typed constant should not validate")
        } catch (expected: IllegalStateException) {
            assertThat<String>(
                    expected.message,
                    containsString("Expected a value of type double, but got string"))
        }

    }

    @Test
    fun doubleWithNonConstantIdentifier() {
        try {
            Constant.validate(symbolTable, IdentifierValueElement(loc, "someStruct", "someStruct"), BuiltinType.DOUBLE)
            fail("Non-constant identifier should not validate")
        } catch (expected: IllegalStateException) {
            assertThat<String>(
                    expected.message,
                    containsString("Unrecognized const identifier: someStruct"))
        }

    }

    @Test
    fun enumWithMember() {
        val memberElements = listOf(EnumMemberElement(loc, "TEST", 1))

        val enumElement = EnumElement(loc, "TestEnum", memberElements)

        val et = EnumType(enumElement, emptyMap())

        Constant.validate(symbolTable, IdentifierValueElement(loc, "TestEnum.TEST", "TestEnum.TEST"), et)
    }

    @Test
    fun enumWithNonMemberIdentifier() {
        val memberElements = listOf(EnumMemberElement(loc, "TEST", 1))

        val enumElement = EnumElement(loc, "TestEnum", memberElements)

        val et = EnumType(enumElement, emptyMap())

        try {
            Constant.validate(symbolTable, IdentifierValueElement(loc, "TestEnum.NON_MEMBER", "TestEnum.NON_MEMBER"), et)
            fail("Non-member identifier should fail")
        } catch (expected: IllegalStateException) {
            assertThat<String>(
                    expected.message,
                    containsString("'TestEnum.NON_MEMBER' is not a member of enum type TestEnum: members=[TEST]"))
        }

    }

    @Test
    fun unqualifiedEnumMember() {
        val memberElements = listOf(EnumMemberElement(loc, "TEST", 1))

        val enumElement = EnumElement(loc, "TestEnum", memberElements)

        val et = EnumType(enumElement, emptyMap())

        try {
            Constant.validate(symbolTable, IdentifierValueElement(loc, "TEST", "TEST"), et)
            fail("Expected an IllegalStateException")
        } catch (e: IllegalStateException) {
            assertThat<String>(e.message, containsString("Unqualified name 'TEST' is not a valid enum constant value"))
        }

    }

    @Test
    fun listOfInts() {
        val list = ListType(BuiltinType.I32)
        val listValue = ListValueElement(location = loc, thriftText = "[0, 1, 2]", value = listOf(
                IntValueElement(loc, "0", 0),
                IntValueElement(loc, "1", 1),
                IntValueElement(loc, "2", 2)
        ))

        Constant.validate(symbolTable, listValue, list)
    }

    @Test
    fun heterogeneousList() {
        val list = ListType(BuiltinType.I32)
        val listValue = ListValueElement(location = loc, thriftText = "[0, 1, \"2\"]", value = listOf(
                IntValueElement(loc, "0", 0),
                IntValueElement(loc, "1", 1),
                LiteralValueElement(loc, "\"2\"", "2")
        ))

        try {
            Constant.validate(symbolTable, listValue, list)
            fail("Heterogeneous lists should fail validation")
        } catch (ignored: IllegalStateException) {
        }

    }

    @Test
    fun typedefOfEnum() {
        val memberElements = listOf(EnumMemberElement(loc, "FOO", 1))

        val enumElement = EnumElement(loc, "AnEnum", memberElements)

        val et = EnumType(enumElement, emptyMap())

        val typedefType = makeTypedef(et, "Id")

        val value = IdentifierValueElement(loc, "AnEnum.FOO", "AnEnum.FOO")

        Constant.validate(symbolTable, value, typedefType)
    }

    @Test
    fun typedefOfWrongEnum() {
        val wrongMemberElements = listOf(EnumMemberElement(loc, "BAR", 2))

        val wrongEnumElement = EnumElement(loc, "DifferentEnum", wrongMemberElements)

        val wt = EnumType(wrongEnumElement, emptyMap())

        val typedefType = makeTypedef(wt, "Id")

        val value = IdentifierValueElement(loc, "AnEnum.FOO", "AnEnum.FOO")

        try {
            Constant.validate(symbolTable, value, typedefType)
            fail("An enum literal of type A cannot be assigned to a typedef of type B")
        } catch (expected: IllegalStateException) {
            assertThat<String>(
                    expected.message,
                    `is`("'AnEnum.FOO' is not a member of enum type DifferentEnum: members=[BAR]"))
        }

    }

    @Test
    fun setOfInts() {
        val setType = SetType(BuiltinType.I32)
        val listValue = ListValueElement(location = loc, thriftText = "[0, 1, 2]", value = listOf(
                IntValueElement(loc, "0", 0),
                IntValueElement(loc, "1", 1),
                IntValueElement(loc, "2", 2)
        ))

        Constant.validate(symbolTable, listValue, setType)
    }

    private fun makeConstant(name: String, typeElement: TypeElement, value: ConstValueElement, thriftType: ThriftType): Constant {
        val element = ConstElement(
                loc,
                typeElement,
                name,
                value)

        return Constant(element, emptyMap(), thriftType)
    }

    private fun makeTypedef(oldType: ThriftType, newName: String): TypedefType {
        val element = TypedefElement(
                loc,
                ScalarTypeElement(loc, "does_not_matter", null),
                newName)

        return TypedefType(element, emptyMap(), oldType)
    }
}
